kernel/src/process.rs

```diff
@@ -1143,43 +1290,190 @@ impl<C: Chip> ProcessType for Process<'_, C> {
             })
     }
 
-    fn allow(
+    #[allow(clippy::not_unsafe_ptr_arg_deref)]
+    fn allow_readwrite(
+        &self,
+        buf_start_addr: *mut u8,
+        size: usize,
+        driver_invoc_closure: &dyn Fn(
+            ReadWriteAppSlice,
+        )
+            -> Result<ReadWriteAppSlice, (ReadWriteAppSlice, ErrorCode)>,
+    ) -> GenericSyscallReturnValue {
+        if !self.is_active() {
+            // Do not operate on an inactive process
+            return GenericSyscallReturnValue::AllowReadWriteFailure(
+                ErrorCode::FAIL,
+                buf_start_addr,
+                size,
+            );
+        }
+
+        // A process is allowed to pass any pointer if the slice
+        // length is 0, as to revoke kernel access to a memory region
+        // without granting access to another one
+        let new_slice = if size == 0 {
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as a slice
+            // of length 0 will never allow dereferencing any memory
+            // in a safe manner.
+            unsafe { ReadWriteAppSlice::new(buf_start_addr, 0, self.appid()) }
+        } else if self.in_app_owned_memory(buf_start_addr, size) {
+            // TODO: Check for buffer aliasing here
```

Comment:
Todo? This seems important from an undefined behavior point of view?

Reply:
Yes. This is very important. But I think we agreed that enforcing kernel invariants would be post-merge, pre-release? There are a few implementation options here which we can discuss.

Reply:
Yes. This is an important invariant to enforce, however my current sketched out version of this shares a lot of the core ideas behind the callback swap prevention (especially the current approach of #2462), most of which could also be used for a more efficient implementation of anti buffer-aliasing for AppSlices.

Reply:
marking resolved as this will be handled post-merge